streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java

```diff
@@ -260,20 +265,16 @@ public void close() {
     private void shutdown() {
         log.info("stream-thread [{}] Shutting down", this.getName());
 
-        // Exceptions should not prevent this call from going through all shutdown steps
-        try {
-            commitAll();
-        } catch (Throwable e) {
-            // already logged in commitAll()
-        }
-
-        // Close standby tasks before closing the restore consumer since closing standby tasks uses the restore consumer.
-        removeStandbyTasks();
+        // Commit first as there may be cached records that have not been flushed yet.
+        commitQuietly();
```

Comment:
There are some redundant work in `commit` on line 269, as it also includes flushing the producer, which takes time (it blocks on all records to be acked from the broker); whereas we only need to `explicitly flush the stores` here.

At the same time, when `closing the tasks` at line 271, new records could be written to the state stores, and maybe cached. So after closing the tasks the state stores need to be flushed again.

So how about using the `state manager flushing` to distinguish from `closing` of state stores in the state managers (now they are done together in  `stateMgr.close`), and then re-order the shutting down process as

1. Commit tasks BUT only commit their consumers, do not flush states or flush producers. 
2. Close tasks.
3. Flush stores: `stateMgr.flush(processorContext)`.
4. Flush producers.
5. Close state manager: `closeAllStateManagers()`.
6. Close producer / consumer / restore consumer.